[JsonPath] Add FrameworkBundle integration

alexandre-daubois opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.3
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Issues        | -
| License       | MIT

First, I'd like to rename `JsonCrawlerInterface` to `CrawlerInterface`. As stated in the early feedback of the PR, the parsed data explored with the JSONPath syntax could be various (explained in the first lines of https://github.com/symfony/symfony/pull/60083#issuecomment-2764246214). It seems better to remove `Json` from `CrawlerInterface`, the latter describing an interface capable of crawling arbitrary data structures with JsonPath.

Also, JsonCrawler now takes data from the `withRawData()` method. What's the idea behind the factory? The JSONPath RFC allows you to create new custom functions to filter your results. I'd like to create a new `#[AsJsonPathFunction]` attribute allowing to register custom functions. To have the best DX, we'll probably want to inject a `JsonCrawler\CrawlerInterface` in our services with custom functions already registered thanks to the attribute and a container call to a future `withCustomFunctions()` method. This is currently not possible this prototype method because the `JsonCrawler` object is stateful.

I didn't include a CHANGELOG line as the component is not yet released. Does it need one?

GromNaN commented on src/Symfony/Component/JsonPath/Tests/JsonCrawlerTest.php:
```diff
@@ -25,20 +25,21 @@ public function testNotStringOrResourceThrows()
         $this->expectExceptionMessage('Expected string or resource, got "int".');
 
-        new JsonCrawler(42);
+        $crawler = new JsonCrawler();
```
The exception is expected in the `find` method, I would move the `new JsonCrawler` before `expectException`. So that we are more specific in which operation causes the error.

alexandre-daubois replied:
Indeed!

GromNaN commented on src/Symfony/Component/JsonPath/JsonCrawler.php:
```diff
@@ -38,36 +38,29 @@ final class JsonCrawler implements JsonCrawlerInterface
+    public function find(string|JsonPath $query, mixed $data): array
+    {
+        if (!\is_string($data) && !\is_resource($data)) {
+            throw new InvalidArgumentException(\sprintf('Expected string or resource, got "%s".', get_debug_type($data)));
```
This exception should be a [`TypeError`](https://www.php.net/manual/en/class.typeerror.php). 
> The argument type being passed to a function does not match its corresponding declared parameter type. 

The type is declared using an annotation.
```suggestion
            throw new \TypeError(\sprintf('Expected string or resource, got "%s".', get_debug_type($data)));
```

alexandre-daubois replied:
I updated this part, good point. Thanks!

GromNaN commented on src/Symfony/Component/JsonPath/JsonCrawlerInterface.php:
```diff
@@ -22,10 +22,12 @@
      * @throws JsonCrawlerException When a syntax error occurs in the provided JSON path
      */
-    public function find(string|JsonPath $query): array;
+    public function find(string|JsonPath $query, mixed $data): array;
```
How did do you decide of the order of the arguments?
I think we should put `$data` before `$query` to be consistent with string methods that put the data (haystack) before the query (needle)

alexandre-daubois replied:
It sticks to array functions order, because in the end, it allows to search in collections. But both orders suit me fine honestly 🙂

chalasr replied:
I tend to put loosely-typed parameters last, but Jérôme's reasoning sounds good

GromNaN commented:
On second thought, maybe a factory is what is needed, to split the responsability between reading the source and executing a JSON path query.

I prefer `JsonCrawlerInterface` as it was, because it doesn't constrain the format of the JSON input. If you want to make an implementation that reads nested arrays or even objects (with PropertyAccessor?), it's possible. When the data is an argument to the `find` method, that restricts the way it can work.

So I suggest a `JsonCrawlerFactory` class used as a service, with a method (`parse`?) that takes the JSON and returns a JsonCrawler instance, injecting the functions. 

```php
class JsonCrawlerFactory implements JsonCrawlerFactoryInterface
{
    public function parse(string|resource $data): JsonCrawlerInterface
    {
        $crawler = new JsonCrawler($data, $this->functions);
    }
}
```

alexandre-daubois commented:
If we want to be consistent and imagine that the input format can be a nested array, maybe the name `JsonCrawlerInterface` isn't right (since we're no longer always crawling a JSON string).

I like the factory idea, however I'm a bit conflicted. If we just look at the current interface without the implementation, I don't find it particularly clear where the data comes from, and this necessarily implies a statefull implementation (with all the drawbacks that implies). 

Maybe we shouldn't put the annotation for `$data` on the interface but only on our implementation? Implementations would be free to check whatever type they want in their logic then. As stated above, this could mean that we should rename the interface to `CrawlerInterface`, defining a class capable to use JsonPath syntax to query any type of data if someone wants to.

alexandre-daubois committed:
[JsonPath] Add FrameworkBundle integration

alexandre-daubois commented:
Thank you for this complete feedback! I just implemented what we discussed. The factory is a bit empty for now, but this will be our extension point as soon as I have some time to implement custom functions 🙂 Feedback welcomed about the factory method's name

nicolas-grekas reviewed (commented):
Please add the FWB integration in the same go;

nicolas-grekas commented on src/Symfony/Component/JsonPath/CrawlerFactoryInterface.php:
```diff
@@ -0,0 +1,25 @@
+    /**
+     * @param string|resource $data
+     */
+    public function createFromData(mixed $data): CrawlerInterface;
```
maybe just create()?

nicolas-grekas replied:
or crawl()?

nicolas-grekas reviewed (commented):
I'm not happy with the factory vocabulary and design pattern.
What about using the prototype pattern instead? Aka wither methods?
Then we'd just need one class, that could be used to derivate configured crawlers.

alexandre-daubois commented:
I reworked the PR to remove the factory. I added a `withRawData()`, so `CrawlerInterface` becomes injectable. Is it better to you?

Open for suggestions about the naming. I noted the [suggestion here](https://github.com/symfony/symfony/pull/60083#discussion_r2042073471), but nor `create` or `crawl` seems right here: we don't create the crawler, and we don't crawl (yet) the JSON string/resource at this point.

OskarStark commented:
Initialize?

alexandre-daubois commented:
@OskarStark why not, but I'm not sure it's a perfect match: wither methods can be called multiple times, where calling "initialize" multiple times feels weird. Maybe `initializeWith()`?

nicolas-grekas commented on src/Symfony/Component/JsonPath/JsonCrawler.php:
```diff
@@ -38,19 +38,34 @@ final class JsonCrawler implements JsonCrawlerInterface
-    public function __construct(
-        private readonly mixed $raw,
-    ) {
+    public function withRawData(mixed $raw): self
```
what about this?

```suggestion
    public function fromJsonString($raw): self
```

alexandre-daubois replied:
I was worried it might be a little too close to the vocabulary of factories, but I'm fine with this name

alexandre-daubois replied:
Actually, `$raw` can be either a string or a resource, so `fromJsonString` is not perfectly accurate. We could either find a better name, or having both `fromJsonString()` and `fromJsonResource()`

nicolas-grekas replied:
It's still accurate to me: the input remains a string, whether it's presented as an actual string or a stream

Which BTW makes me wonder: can one use the JsonPath against an associative array?

alexandre-daubois replied:
> Which BTW makes me wonder: can one use the JsonPath against an associative array?

Currently no, but the support could be fairly easy to bring: `JsonCrawler` only `json_decode()` the raw data then the path is applied on the decoded array. So yes, by providing an array and skipping the decoding step, the whole logic remains the same.

nicolas-grekas replied:
wouldn't it make sense then to support `fromJson(string|array|resource)`?

alexandre-daubois replied:
I don't know. I find it a bit strange to have `fromJson()` which can directly take an array. But at the same time, it's technically so “quick” to add that it makes me doubt. 🙂 

I'm not sure I have a case in mind where someone would give a directly decoded array rather than the JSON string? Especially since, thanks to JsonStreamer, we're able to optimize the part of the string/resource to be decoded. Do you maybe have a use case in mind or is more out of curiosity?
Maybe a second implementation of CrawlerInterface could be added to support this different data structure actually (like Jérôme mentioned with BSON, etc.).

nicolas-grekas replied:
Parsing a response from an API request done with HttpClient for example.

alexandre-daubois replied:
Indeed, in this case I thought the person would rather pass on the raw answer, but I imagine it's a nice bonus for DX anyway without any maintenance cost. I've added support for `array`. Thanks 👍

nicolas-grekas reviewed (commented):
Something doesn't work here:
- if the method isn't in the interface, then there's no point in adding it, since autowiring won't allow using it if proper IoC is used
- then, if we add it to the interface, the input format has to be a JSON string/stream, which invalidates the reasoning that lead to the renaming of the interface

So to me, if we want to make the interface an autowirable type, it has to be named JsonCrawlerInterface, and the method has to be part of it.

Or did I miss anything?

OR, we need two interfaces; JsonCrawlerInterface extends CrawlerInterface :thinking:

alexandre-daubois commented:
I think we should have both interface, that's a nice idea. This would solve the point you raise and still allow Jérôme's [idea](https://github.com/symfony/symfony/pull/60083#issuecomment-2764246214).

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/composer.json:
```diff
@@ -49,6 +49,8 @@
         "symfony/expression-language": "^6.4|^7.0",
         "symfony/html-sanitizer": "^6.4|^7.0",
         "symfony/http-client": "^6.4|^7.0",
+        "symfony/json-path": "^7.3",
```
should be 7.3.* because the component is experimental

nicolas-grekas commented on src/Symfony/Component/JsonPath/JsonCrawler.php:
```diff
@@ -39,18 +39,33 @@ final class JsonCrawler implements JsonCrawlerInterface
+     *
+     * @throws \TypeError When the input is not a string or a resource
+     */
+    public function fromJson(mixed $raw): static
```
```suggestion
    public function fromJson(mixed $data): static
```

nicolas-grekas commented on src/Symfony/Component/JsonPath/JsonCrawler.php:
```diff
@@ -39,18 +39,33 @@ final class JsonCrawler implements JsonCrawlerInterface
+    /**
+     * @param resource|string|array $raw
+     *
+     * @throws \TypeError When the input is not a string or a resource
```
shouldn't be documented

nicolas-grekas commented on src/Symfony/Component/JsonPath/JsonCrawler.php:
```diff
@@ -39,18 +39,33 @@ final class JsonCrawler implements JsonCrawlerInterface
+    public function fromJson(mixed $raw): static
+    {
+        if (!\is_string($raw) && !\is_resource($raw) && !\is_array($raw)) {
+            throw new \TypeError(\sprintf('Expected string, array or resource, got "%s".', get_debug_type($raw)));
```
the error should be explicit about the location, please take inspiration from native type errors

nicolas-grekas commented on src/Symfony/Component/JsonPath/JsonCrawler.php:
```diff
@@ -39,18 +39,33 @@ final class JsonCrawler implements JsonCrawlerInterface
 
     public function find(string|JsonPath $query): array
     {
+        if (null === $this->raw) {
```
let's add a constructor to not have to deal with this case?

alexandre-daubois replied:
Maybe we can use `[]` as the default value of `$raw` instead of `null`: we wouldn't have to deal with this case and the crawler would be noop (i.e. always return `[]`) until one calls `fromJson()`, without adding a constructor. What do you think?

nicolas-grekas replied:
Why not add a constructor? Related changes in the tests look like boilerplate we'd better remove to me.

nicolas-grekas commented on src/Symfony/Component/JsonPath/JsonCrawlerInterface.php:
```diff
@@ -11,21 +11,12 @@
+     * @throws \TypeError When the input is not a string or a resource
      */
-    public function find(string|JsonPath $query): array;
+    public function fromJson(mixed $raw): static;
```
```suggestion
    public function fromJson(mixed $data): static;
```

stof commented:
The suggested JsonCrawlerInterface looks weird to me.

Using it means you will have a property in your class holding a JsonCrawlerInterface on which you should never call the `find()` method (as it would not have any data to actually find something). And you would call `$crawler = $this->crawler->fromJson($data)` to get a working crawler.
To me, the factory being a separate class was making the code a lot more understandable.

nicolas-grekas commented:
Then maybe the too interfaces should be split:
- JsonCrawlerInterface with only have a fromJson(): CrawlerInterface - that's the factory, but named in a more DX-friendly way
- CrawlerInterface, with the find method (and NOT autowirable)

alexandre-daubois commented:
Thanks, I updated the PR to have two interfaces.

Fabbot complains about the TypeError message format. I'd say it is not valid, PHP internal type errors doesn't include type quoting.

nicolas-grekas commented on src/Symfony/Component/JsonPath/JsonCrawler.php:
```diff
@@ -39,16 +39,20 @@ final class JsonCrawler implements JsonCrawlerInterface
         }
     }
 
+    public function fromJson(mixed $data): CrawlerInterface
```
```suggestion
    /**
     * @param resource|string|array $data
     */
    public function fromJson(mixed $data): CrawlerInterface
```

alexandre-daubois replied:
The PHP doc is already present in the interface, should we still add it in the implementing class?

nicolas-grekas replied:
covariance means this could really be widened by the implementation

alexandre-daubois replied:
Oh right, `mixed` is a bit special indeed. Comment added

xabbuh commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -474,6 +475,10 @@ public function load(array $configs, ContainerBuilder $container): void
         }
 
+        if ($this->readConfigEnabled('json_path', $container, $config['json_path'])) {
+            $this->registerJsonPathConfiguration($config['json_streamer'], $container, $loader);
```
```suggestion
            $this->registerJsonPathConfiguration($config['json_path'], $container, $loader);
```

alexandre-daubois replied:
Thanks!

xabbuh commented on src/Symfony/Component/JsonPath/JsonCrawlerInterface.php:
```diff
@@ -22,10 +19,7 @@
 interface JsonCrawlerInterface
```
Do we still need this interface?

alexandre-daubois replied:
Yes, this is our factory and the interface used to be used in service injection.

One of the next step is adding something like `withCustomFunctions()` that gets a tagged iterator of invokable classes to use in JsonPath filters (the RFC allows custom functions in filters).

xabbuh replied:
I wonder though if that doesn't mean that we should rather split the crawler and it's factory. Currently both concerned are mixed in one class which also means that the crawler needs to allow being constructed without data while that shouldn't be needed if the factory was a different class, right?

alexandre-daubois replied:
Indeed. We discussed a bit about it with @nicolas-grekas. We agreed that we're fine having everything in one class as well as having a no-op crawler when created without data (i.e. returning an empty array for all paths). Is there a limitation that you see with this implementation maybe?

I went for the separate factory class at first, but I actually like the fact that if you need a crawler in your service, you can inject `JsonCrawlerInterface` directly instead of having a factory that then creates the crawler.

GromNaN replied:
Having the 2 interfaces implemented by the same class is not a problem when using interfaces for typing.
But I'm not convinced by the name of `JsonCrawlerInterface`.

I would prefer `CrawlerFromJsonInterface` which reuses the method name and gives the intention.

nicolas-grekas replied:
I prefer JsonCrawlerInterface - it feels less convoluted to me, and is explicit about the intended use case - querying json documents.

I also prefer one class to implement the two interfaces. Optimizing the constructor doesn't look worth the split to me.

xabbuh replied:
My main concern with mixing responsibilities in one class is that you could have a crawler in an intermediate state while building and then accidentally pass around the wrong crawler instance which would then behave differently than what you did expect. By separating the crawler and its factory (or probably rather a builder) PHP's type system would save you from such a mistake.

alexandre-daubois replied:
Back from holidays 🙂

I get what you mean. @nicolas-grekas would you be fine introducing the factory which seem backed up in several comments, or do you want to avoid it at all costs?

xabbuh commented on src/Symfony/Component/JsonPath/JsonCrawler.php:
```diff
@@ -39,16 +39,23 @@ final class JsonCrawler implements JsonCrawlerInterface
+    /**
+     * @param resource|string|array $data
+     */
+    public function fromJson(mixed $data): CrawlerInterface
```
What about making the method static? And I would question if we really need an interface for it? And if so, if it's name wouldn't better be `JsonCrawlerFactoryInterface`.

alexandre-daubois replied:
Making the method static wouldn't allow us to plug properly with DI and allow further steps like [this one](https://github.com/symfony/symfony/pull/60083#discussion_r2055527025).

The "factory" name is not used per [this comment](https://github.com/symfony/symfony/pull/60083#pullrequestreview-2775213194) and [this one](https://github.com/symfony/symfony/pull/60083#issuecomment-2821207776).

mtarld replied:
AFAIK, we can have [static factories](https://symfony.com/doc/current/service_container/factories.html#static-factories) in the DI.

But my main question here is about the aim of that method. In fact, such code will look weird to me:
```php
$foo = (new JsonCrawler('"something"'))
    ->fromJson('"otherThing"');
```

Instead, static method looks more logical to me (and still allows you to use the builder pattern):
```php
$foo = JsonCrawler::fromJson('"otherThing"')
    >withCustomFunctions(...);
```

alexandre-daubois replied:
In your last snippet, you call `withCustomFunctions()` yourself, but the goal is to have everything wired correctly so custom functions are automatically registered with an attribute, and you "never" call the function yourself if you use FWB. I'm not sure we'd be able to achieve this with static factories (or any static behavior being used)?

mtarld replied:
You might be able to combine static factories and [service configurators](https://symfony.com/doc/current/service_container/configurators.html) maybe?

nicolas-grekas replied:
Hell no, not configurators.
A static factory would miss the point of configuring transformer functions.

But all this makes me think we're thinking out of the blue. Let's not do anything until we actually do have such transformers.

If we have some lessons to take from this PR to prepare the next steps, let's keep that, but let's remove the autowiring/DI part to the moment where we'll really need it.

mtarld commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/json_path.php:
```diff
@@ -0,0 +1,21 @@
+use Symfony\Component\JsonPath\JsonCrawler;
+use Symfony\Component\JsonPath\JsonCrawlerInterface;
+
+return static function (ContainerConfigurator $container) {
```
Maybe we can add a functional test to make sure that the services are properly registered and usable? (You can find such tests in `Symfony\Bundle\FrameworkBundle\Tests\Functional`

alexandre-daubois commented:
Discussions makes me wonder: is there a design flaw maybe in the current state of the crawler? The number of comments on the PR makes me wonder so.
I took a look at ExpressionLanguage. We could take inspiration from it maybe. ExpressionLanguage takes providers (`ExpressionFunctionProviderInterface[]`) through the constructor.

This could be the same for JsonCrawler, with an `array $functions` or something similar provided through the constructor. Then we could also change the current `JsonCrawler::find($path)` so it accepts the data directly, changing the signature to `JsonPath::find($path, mixed $data)` (args order can be discussed 🙂). We can keep an interface describing the `find()` method, this is a nice extension point as pointed by Jérôme.

Status: Needs work
